### ISSUE TITLE
array display is a mess

### ISSUE THREAD

**Issue by StefanKarpinski** [JuliaLang/julia#6117]
Recent changes to array printing have made it impossible to normally provide show methods for custom array types. The writemime function intercepts printing of arrays, so that if you define a new type, e.g. `Bytes <: AbstractVector{Uint8}` and then define `show(io::IO, b::Bytes)`, that show method never gets called. This whole situation is a pretty huge mess, imo. There's no good reason for show and writemime for arrays to produce such different results and you certainly shouldn't have to know about or override writemime methods to get a custom array type to show up the way you want it to in the repl.

**Comment by diegozea** [comment]
I have found this problem when I was defining an `AbstractArray`. I solved it with a new definition of `Base.showarray` for my type. Is there another way to do this? Should this be documented in Abstract Array Interface?
```
julia> N = zeros(ResidueCount{Int, 2, false})
20x20 ResidueCount{Int64,2,false}:
Error showing value of type ResidueCount{Int64,2,false}:
ERROR: MethodError: `ndims` has no method matching ndims(::ResidueCount{Int64,2,false})
 in showarray at show.jl:1205
 in anonymous at replutil.jl:29
 in with_output_limit at ./show.jl:1242
 in writemime at replutil.jl:28
 in display at REPL.jl:113
 in display at REPL.jl:116
 in display at multimedia.jl:151
 in print_response at REPL.jl:133
 in print_response at REPL.jl:120
 in anonymous at REPL.jl:621
 in run_interface at ./LineEdit.jl:1586
 in run_frontend at ./REPL.jl:862
 in run_repl at ./REPL.jl:165
 in _start at ./client.jl:453

julia> show(N)
ResidueCount{Int64,2,false} 
  counts: Array(Int64,(20,20)) [0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0]
  marginals: Array(Int64,(2,20)) [0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0]
  total: Int64 0
```

**Comment by mbauman** [comment]
It's a mess, but it shouldn't be erroring.  What's the definition of `ResidueCount`?  You need to pass your `eltype` and `ndims` parameters to `AbstractArray`, e.g.,

```jl
immutable ResidueCount{T, N, b} <: AbstractArray{T,N}
```

**Comment by lindahua** [comment]
We came across similar problems as to how to properly specialize the `show` of an array. 

I think we need a specific page in the manual explaining how to create a new array type (e.g. what methods need to be defined, and what their behaviors are supposed to be, etc).

**Comment by diegozea** [comment]
Thanks @mbauman, I wasn't passing my parameters to `AbstractArray`, now I'm getting:
```
julia> N = zeros(ResidueCount{Int, 2, false})
20x20 ResidueCount{Int64,2,false}:
 #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef
 #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef
 #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef
 #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef
 #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef
 #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef
 #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef
 #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef
 #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef
 #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef
 #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef
 #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef
 #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef
 #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef
 #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef
 #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef
 #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef
 #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef
 #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef
 #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef  #undef

julia> N[1,1]
ERROR: indexing not defined for ResidueCount{Int64,2,false}
 in getindex at abstractarray.jl:459

julia> Base.getindex(n::ResidueCount, i, j) = getindex(n.counts, i, j)
getindex (generic function with 126 methods)

julia> N[1,1]
0

julia> N = zeros(ResidueCount{Int, 2, false})
20x20 ResidueCount{Int64,2,false}:
Error showing value of type ResidueCount{Int64,2,false}:
ERROR: indexing not defined for ResidueCount{Int64,2,false}
 in alignment at show.jl:967
 in print_matrix at show.jl:1042
 in showarray at show.jl:1225
 in anonymous at replutil.jl:29
 in with_output_limit at ./show.jl:1242
 in writemime at replutil.jl:28
 in display at REPL.jl:113
 in display at REPL.jl:116
 in display at multimedia.jl:151
 in print_response at REPL.jl:133
 in print_response at REPL.jl:120
 in anonymous at REPL.jl:621
 in run_interface at ./LineEdit.jl:1586
 in run_frontend at ./REPL.jl:862
 in run_repl at ./REPL.jl:165
 in _start at ./client.jl:453

```
Why am I getting `#undef`? My `getindex` definitions are:
```
Base.linearindexing(ResidueCount)	= Base.LinearFast()

getindex(n::ResidueCount, i::Int)	= getindex(n.counts, i)
```

**Comment by IainNZ** [comment]
@lindahua that exists: http://docs.julialang.org/en/latest/manual/interfaces/#abstract-arrays

**Comment by lindahua** [comment]
Thanks for the pointer. But that part of the document does not mention `showarray` and other printing related stuff.

**Comment by IainNZ** [comment]
Correct.

**Comment by vtjnash** [comment]
see #14052 for latest cleanup plans for this

**Comment by JeffBezanson** [comment]
Fixed by #16354.
